src/Components/HistoricalDataChart/Util.js

```diff
@@ -0,0 +1,61 @@
+export const formatHistoricalData = (responseData, dateRange) =>{
+
+    const chartData = [];
+    let today = new Date();
+
+    const cpu = {
+        datapoints: []
+    };
+
+    const memory = {
+        datapoints: []
+    };
+
+    let dataFound = false;
+
+    for (let i = (dateRange - 1); i >= 0; i--) {
+        dataFound = false;
+        let dateToCheck = new Date();
+        dateToCheck =   new Date(dateToCheck.setDate(today.getDate() - i)).toDateString();
+
+        responseData.map((dataObj) => {
+            let reportDate = new Date(dataObj.report_date).toDateString();
+
+            if (!dataFound && reportDate === dateToCheck) {
+
+                cpu.datapoints.push({
+                    name: 'CPU Utilization',
+                    x: new Date(dateToCheck),
+                    y: dataObj.cpu
+                });
+
+                memory.datapoints.push({
+                    name: 'Memory Utilization',
+                    x: new Date(dateToCheck),
+                    y: dataObj.memory
+                });
+
+                dataFound = true;
+            }
+        });
+
+        if (!dataFound) {
+            cpu.datapoints.push({
+                name: 'CPU Utilization',
+                x: new Date(dateToCheck),
+                y: null
+            });
+
+            memory.datapoints.push({
+                name: 'Memory Utilization',
+                x: new Date(dateToCheck),
+                y: null
+            });
+        }
+    }
```

Comment:
Hello @r14chandra the reason I made this change is because the backend respnse only contains records where data is available and for chart we need to create data points for entire 7/45 days and hence I had to write this map to go through response for each date and extract record if its available or set it to null if value is not there. I agree this needs improvement in future and I have discussed the same with @kgaikwad and may be later we can discuss both on frontend/backend side regarding the same. I hope that answer your question.

Reply:
@r14chandra Thank you for suggesting the changes, the above modification really helped in optimizing the code 🙌🏻 
I have committed and push the above changes in the code 👍🏻